### PR TITLE
Add support for arrays of geo_shape

### PR DESCRIPTION
**Pull request by matriv**

### CODE REVIEW

File: sql/src/main/java/io/crate/protocols/postgres/types/JsonInnerType.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    private static final ObjectMapper JSON_MAPPER = new ObjectMapper();
+
+    private JsonInnerType() {
+        super();
+    }
+
+    @Override
+    protected byte[] encodeAsUTF8Text(@Nonnull Object value) {
+        try {
+            String json = JSON_MAPPER.writeValueAsString(value);
```
Comment (dobe) [inline sql/src/main/java/io/crate/protocols/postgres/types/JsonInnerType.java]: try to use xontent stuff and work on bytes

File: sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java [hunk 2]
```diff
@@ -182,11 +188,15 @@ private void addObject(byte[] bytes, int startIdx, int endIdx, List<Object> obje
...
             byte firstValueByte = bytes[startIdx];
             if (firstValueByte == '"') {
-                byte[] innerBytes = Arrays.copyOfRange(bytes, startIdx + 1, endIdx);
-                String s = new String(innerBytes, StandardCharsets.UTF_8);
-                s = s.replace("\\\"", "\"");
-                s = s.replace("\\\\", "\\");
-                objects.add(innerType.decodeUTF8Text(s.getBytes(StandardCharsets.UTF_8)));
+                List<Byte> innerBytes = new ArrayList<>(endIdx - (startIdx + 1));
+                for (int i = startIdx + 1; i < endIdx; i++) {
+                    if (i < (endIdx + 1) && (char)bytes[i] == '\\' &&
```
Comment (dobe) [inline sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java]: i <= endIdx

File: sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java [hunk 3]
```diff
@@ -95,11 +95,17 @@ public Object readBinaryValue(ChannelBuffer buffer, int valueLength) {
                     }
                 } else {
                     bytes = innerType.encodeAsUTF8Text(o);
-                    encodedValues.add((byte) '\"');
+                    boolean isJson = JsonType.OID == innerType.oid();
+
+                    encodedValues.add((byte) '"');
                     for (byte aByte : bytes) {
+                        // Escape double quotes with backslash for json
+                        if (isJson && (char) aByte == '"') {
```
Comment (dobe) [inline sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java]: isJson can be outside of the loop

File: sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java (hunk 2 above)
Comment (matriv) [inline sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java]: The Arrays.copyOfRange that we used before it's excluding the endIdx..

File: sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java (hunk 3 above)
Comment (matriv) [inline sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java]: You mean something like that?
`if (isJson) {
    for (byte aByte : bytes) {
        // Escape double quotes with backslash for json
        if ((char) aByte == '"') {
            encodedValues.add((byte) '\\');
        }
        encodedValues.add(aByte);
    }
} else {
    for (byte aByte : bytes) {
        encodedValues.add(aByte);
    }
}`

File: sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java (hunk 2 above)
Comment (dobe) [inline sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java]: x<y+1 ==  x<=y

File: sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java (hunk 3 above)
Comment (dobe) [inline sql/src/main/java/io/crate/protocols/postgres/types/PGArray.java]: exactly

**Comment by dobe** [comment]
ack

**Pull request merged**